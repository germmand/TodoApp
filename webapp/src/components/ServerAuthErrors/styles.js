export default theme => ({
  root: {
    ...theme.mixins.gutters(),
    paddingTop: theme.spacing.unit * 2,
    paddingBottom: theme.spacing.unit * 2,
    marginBottom: theme.spacing.unit * 2,
    width: 485,
    backgroundColor: theme.palette.danger !== undefined ? theme.palette.danger.main : '#ED4740',
  },
});
