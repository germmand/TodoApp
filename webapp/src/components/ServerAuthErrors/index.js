import React from 'react';
import PropTypes from 'prop-types';

import { withStyles } from '@material-ui/core/styles';
import {
  Paper,
  List,
  ListItem,
  ListItemIcon,
  ListItemText,
  ListSubheader,
} from '@material-ui/core';
import ErrorIcon from '@material-ui/icons/Error';
import styles from './styles';

import capitalizeFirstLetter from '../../common/functions/capitalizeFirstLetter';

const ServerAuthErrors = ({ errors, classes }) => (
    <Paper className={classes.root} elevation={1}>
        <List
            component="nav"
            subheader={
                <ListSubheader component="div">Server responded with some errors:</ListSubheader>
            }
        >
            {Object.entries(errors).map(serverErrors => (
                    <ListItem divider key={serverErrors[0]}>
                        <ListItemIcon>
                            <ErrorIcon />
                        </ListItemIcon>
                        <ListItemText
                            primary={serverErrors[0] === 'non_field_errors'
                              ? 'Credentials'
                              : capitalizeFirstLetter(serverErrors[0])
                            }
                            secondary={serverErrors[1][0]}
                        />
                    </ListItem>
            ))}
        </List>
    </Paper>
);

ServerAuthErrors.propTypes = {
  classes: PropTypes.objectOf(PropTypes.string).isRequired,
  errors: PropTypes.objectOf(PropTypes.array).isRequired,
};

export default withStyles(styles)(ServerAuthErrors);
