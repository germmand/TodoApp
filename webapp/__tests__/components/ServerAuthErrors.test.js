import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import ServerAuthErrors from '../../src/components/ServerAuthErrors';
import capitalizeFirstLetter from '../../src/common/functions/capitalizeFirstLetter';

describe('ServerAuthErrors Component', () => {
  it('renders ServerAuthErrors component correctly', () => {
    const props = {
      classes: {
        root: 'root-container',
      },
      errors: {
        credentials: ['No account was found with given credentials.'],
      },
    };
    const ServerAuthErrorsComponent = renderer.create(<ServerAuthErrors {...props} />);
    expect(ServerAuthErrorsComponent.toJSON()).toMatchSnapshot();
  });

  it('replaces \'non_field_errors\' with \'Credentials\'', () => {
    const props = {
      classes: {
        root: 'root-container',
      },
      errors: {
        non_field_errors: ['No account was found with given credentials.'],
      },
    };
    const ServerAuthErrorsComponent = mount(<ServerAuthErrors {...props} />);
    const ListItemTextComponents = ServerAuthErrorsComponent.find('ListItemText');
    expect(ListItemTextComponents).toHaveLength(1);
    expect(ListItemTextComponents.at(0).props().primary).toEqual('Credentials');
  });

  it('capitalizes first letter on primary errors', () => {
    const props = {
      classes: {
        root: 'root-container',
      },
      errors: {
        username: ['This field can\'t be blank.'],
      },
    };
    const expectedErrorTitle = capitalizeFirstLetter(Object.keys(props.errors)[0]);
    const ServerAuthErrorsComponent = mount(<ServerAuthErrors {...props} />);
    const ListItemTextComponents = ServerAuthErrorsComponent.find('ListItemText');
    expect(ListItemTextComponents).toHaveLength(1);
    expect(ListItemTextComponents.at(0).props().primary).toEqual(expectedErrorTitle);
  });

  it('renders first error description message properly on secondary', () => {
    const props = {
      classes: {
        root: 'root-container',
      },
      errors: {
        username: ['This field can\'t be blank.'],
      },
    };
    const ServerAuthErrorsComponent = mount(<ServerAuthErrors {...props} />);
    const ListItemTextComponents = ServerAuthErrorsComponent.find('ListItemText');
    expect(ListItemTextComponents).toHaveLength(1);
    expect(ListItemTextComponents.at(0).props().secondary).toEqual(props.errors.username[0]);
  });
});
