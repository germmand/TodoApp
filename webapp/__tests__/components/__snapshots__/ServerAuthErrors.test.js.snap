// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ServerAuthErrors Component renders ServerAuthErrors component correctly 1`] = `
<div
  className="MuiPaper-root-2 MuiPaper-elevation1-5 MuiPaper-rounded-3 ServerAuthErrors-root-1 root-container"
>
  <nav
    className="MuiList-root-29 MuiList-padding-30 MuiList-subheader-32"
  >
    <div
      className="MuiListSubheader-root-33 MuiListSubheader-sticky-38 MuiListSubheader-gutters-36"
    >
      Server responded with some errors:
    </div>
    <li
      className="MuiListItem-root-39 MuiListItem-default-42 MuiListItem-gutters-47 MuiListItem-divider-46"
      disabled={false}
    >
      <div
        className="MuiListItemIcon-root-51"
      >
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root-52"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-2h2v2zm0-4h-2V7h2v6z"
          />
        </svg>
      </div>
      <div
        className="MuiListItemText-root-61"
      >
        <span
          className="MuiTypography-root-67 MuiTypography-subheading-74 MuiListItemText-primary-64"
        >
          Credentials
        </span>
        <p
          className="MuiTypography-root-67 MuiTypography-body1-76 MuiTypography-colorTextSecondary-100 MuiListItemText-secondary-65"
        >
          No account was found with given credentials.
        </p>
      </div>
    </li>
  </nav>
</div>
`;
